### PR TITLE
nixos/{jitsi-meet,prosody}: Support JWT authentication in Jitsi

### PR DESCRIPTION
**Pull request by numinit**
###### Motivation for this change

Jitsi doesn't work with JWT authentication and needs in-tree Lua modules added to Prosody.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by numinit** [comment]
Will require updates once #152877 is merged.

**Comment by numinit** [comment]
Apologies for all the commits, I'll rebase once #152877 is merged.

**Comment by numinit** [comment]
Seems we have jitsi-meet-prosody now, but it's just pulling from a deb. Is this method preferable or can I simplify this PR?

**Comment by numinit** [comment]
Should be ready for review as latest Jitsi changes have been merged.

File: pkgs/servers/xmpp/prosody/default.nix [hunk 1]
```diff
@@ -1,12 +1,13 @@
...
 , lua
 , nixosTests
 , withLibevent ? true
 , withDBI ? true
 # use withExtraLibs to add additional dependencies of community modules
 , withExtraLibs ? [ ]
 , withOnlyInstalledCommunityModules ? [ ]
-, withCommunityModules ? [ ] }:
+, withCommunityModules ? (lib.optionals withJitsiModules ["auth_token"])
+, withJitsiModules ? true }:
```
Comment (NinjaTrappeur) [inline pkgs/servers/xmpp/prosody/default.nix:10]:
```suggestion
, withJitsiModules ? false }:
```

File: pkgs/servers/xmpp/prosody/default.nix [hunk 2]
```diff
@@ -47,6 +49,17 @@ stdenv.mkDerivation rec {
   };
 
   nativeBuildInputs = [ makeWrapper ];
+
+  # Be sure to update this whenever jitsi-meet is updated.
+  jitsiMeet = if withJitsiModules then
```
Comment (NinjaTrappeur) [inline pkgs/servers/xmpp/prosody/default.nix:54]: TOREMOVE

Instead, you can copy the `/resources/prosody-plugins` directory to a specific location (eg. `$out/resources/prosody-plugins`) in the `jitsi-meet` derivation.

File: pkgs/servers/xmpp/prosody/default.nix [hunk 3]
```diff
@@ -68,6 +81,7 @@ stdenv.mkDerivation rec {
       ${concatMapStringsSep "\n" (module: ''
         cp -r $communityModules/mod_${module} $out/lib/prosody/modules/
       '') (lib.lists.unique(nixosModuleDeps ++ withCommunityModules ++ withOnlyInstalledCommunityModules))}
+      ${lib.optionalString withJitsiModules "cp -R $jitsiMeet/resources/prosody-plugins/* $out/lib/prosody/modules/"}
```
Comment (NinjaTrappeur) [inline pkgs/servers/xmpp/prosody/default.nix:84]:
```suggestion
      ${lib.optionalString withJitsiModules "cp -R ${jitsi-meet}/resources/prosody-plugins/* $out/lib/prosody/modules/"}
```

File: pkgs/servers/xmpp/prosody/default.nix [hunk 4]
```diff
@@ -1,12 +1,13 @@
-{ stdenv, fetchurl, lib, libidn, openssl, makeWrapper, fetchhg
+{ stdenv, fetchurl, lib, libidn, openssl, makeWrapper, fetchhg, fetchFromGitHub
```
Comment (NinjaTrappeur) [inline pkgs/servers/xmpp/prosody/default.nix:1]:
```suggestion
{ stdenv, fetchurl, lib, libidn, openssl, makeWrapper, fetchhg, fetchFromGitHub, jitsi-meet
```

**Review by picnoir — CHANGES_REQUESTED** [review]
Reviewing the prosody side only:

We should remove any jitsi mention from the default prosody derivation.

As it is, this PR will enable the jitsi-meet extensions to the stock prosody NixOS module.

**Comment by numinit** [comment]
> Reviewing the prosody side only:
> 
> We should remove any jitsi mention from the default prosody derivation.
> 
> As it is, this PR will enable the jitsi-meet extensions to the stock prosody NixOS module.

Thanks. Wasn't sure what the defaults should be, but I agree. Will make these changes.

**Comment by numinit** [comment]
Superseded by #207886.

**Pull request closed without merging**